### PR TITLE
feat: Update keycloak from 6.0.1 to 15.

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### What does this PR do?
Update keycloak from 6.0.1 to 15.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/19585

### Related pr:

https://github.com/eclipse-che/che-operator/pull/1015

## Test installation
1. Build keycloak image for this pull request:

``` bash
  $ cd dockerfiles/keycloak/  && ./build.sh 
```

Retag and push image quay.io/eclipse/che-keycloak:next. **You can use my image**: quay.io/aandriienko/che-keycloak:next-k15

2. Create CR patch to provide custom keycloak image

```
$ touch ~/newKeycloakCR.yaml
```

```
spec:
  auth:
    identityProviderImage: 'quay.io/aandriienko/che-keycloak:next-k15'
```

3. Clone che-operator

```bash
$ git clone git@github.com:eclipse-che/che-operator.git
$ cd che-operator
$ git checkout keycloak-15
```

5. Build custom che-operator image 

```bash
$ make docker-build docker-push IMG="${IMAGE_REGISTRY_HOST}/${IMAGE_REGISTRY_USER_NAME}/che-operator:next-k15"
```

**You can use my image** quay.io/aandriienko/che-operator:next-k15

5. Prepare chectl template folder:

```bash
$ mkdir -p /tmp/templates/che-operator
$ make chectl-templ SRC=~/projects/che-operator TARGET=/tmp/templates/che-operator
```

 Install Che using custom che-operator and keycloak image on the minikube

```bash
chectl server:deploy -n che -p minikube --installer operator --che-operator-cr-patch-yaml=${HOME}/newKeycloakCR.yaml --che-operator-image=quay.io/aandriienko/che-operator:next-k15 --templates=/tmp/templates
```

Notice: to deploy on the Openshift 4, change platform name from `minikube` to `openshift`

## Test update from old version to the newer

1. Deploy stable version Che:

```bash
$ chectl server:deploy -p minikube -n che --version=7.35.0 --installer operator
```

2. Update Che:

```bash
chectl server:update -n che -p minikube --installer operator --che-operator-cr-patch-yaml=${HOME}/newKeycloakCR.yaml --che-operator-image=quay.io/aandriienko/che-operator:next-k15 --templates=/tmp/templates -y
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

Signed-off-by: Oleksandr Andriienko <oandriie@redhat.com>

### CODE REVIEW

File: dockerfiles/keycloak/Dockerfile [hunk 1]
```diff
@@ -6,15 +6,15 @@
...
 ADD che /opt/jboss/keycloak/themes/che
 ADD che-username-readonly /opt/jboss/keycloak/themes/che-username-readonly
 ADD . /scripts/
 ADD cli /scripts/cli
-RUN ln -s /opt/jboss/tools/docker-entrypoint.sh && \
-    curl -sSL https://github.com/che-incubator/KEYCLOAK-10169-OpenShift4-User-Provider/releases/download/6.0.1-openshift-v4/openshift4-extension-6.0.1.jar -o /opt/jboss/keycloak/standalone/deployments/openshift4-extension-6.0.1.jar && \
-    unzip -j /opt/jboss/keycloak/standalone/deployments/openshift4-extension-6.0.1.jar -d /opt/jboss/keycloak/themes/base/admin/resources/partials \
-    theme-resources/resources/realm-identity-provider-openshift-v4.html theme-resources/resources/realm-identity-provider-openshift-v4-ext.html
+USER root
+RUN microdnf install findutils nano && microdnf clean all; \
```
Comment (tolusha) [inline dockerfiles/keycloak/Dockerfile:16]: Should we use `&&` instead of `;` ?

Comment (AndrienkoAleksandr) [reply dockerfiles/keycloak/Dockerfile:16]: they should work both, but ok I will use &&

Comment (benoitf) [reply dockerfiles/keycloak/Dockerfile:16]: it's not clear why we would require nano and findutils ?

File: dockerfiles/keycloak/Dockerfile [hunk 2]
```diff
@@ -6,15 +6,15 @@
 # SPDX-License-Identifier: EPL-2.0
 #
 
-FROM jboss/keycloak:6.0.1
+FROM jboss/keycloak:15.0.0
```
Comment (benoitf) [inline dockerfiles/keycloak/Dockerfile:9]: Can we use quay.io/keycloak/keycloak ?

Comment (AndrienkoAleksandr) [reply dockerfiles/keycloak/Dockerfile:9]: Yes we can.

File: dockerfiles/keycloak/Dockerfile (hunk 1 above)
Comment (AndrienkoAleksandr) [reply dockerfiles/keycloak/Dockerfile:16]: Ops, nano I will remove. findutils we need for che-operator. Some settings which we configure with help of che-operator need `find` command

File: assembly/assembly-root-war/src/main/webapp/_app/keycloackLoader.js [hunk 3]
```diff
@@ -110,8 +110,8 @@ export class KeycloakLoader {
                     onLoad: 'login-required',
                     checkLoginIframe: false,
                     useNonce: useNonce,
-                    scope: 'email profile',
-                    redirectUri: keycloakSettings['che.keycloak.redirect_url.ide']
+                    scope: 'openid',
```
Comment (skabashnyuk) [inline assembly/assembly-root-war/src/main/webapp/_app/keycloackLoader.js:113]: Can you explain the changes in scope. Why it is needed? Is there a change in JWT token content?

Comment (AndrienkoAleksandr) [reply assembly/assembly-root-war/src/main/webapp/_app/keycloackLoader.js:113]: I don't think, that I changed scope. Keycloak 6.0.1 adapters ignored this option from workspace loader and set up `openid` scope by default. New keycloak adapters respect this option. So I only set up the same value like it actually was before. Also in the get token response with old and new keycloak I see in the json scope field: 'openid email profile'. When I decode jwt token from new keycloak I really see fields:

...
  "scope": "openid email profile",
  "sid": "4d9e49e6-b648-40e9-bd63-eea3d99408e5",
  "email_verified": false,
  "name": "Admin Admin",
  "preferred_username": "admin",
  "given_name": "Admin",
  "family_name": "Admin",
  "email": "admin@admin.com"
...

So I don't see any fundamental difference.

Comment (skabashnyuk) [reply assembly/assembly-root-war/src/main/webapp/_app/keycloackLoader.js:113]: thanks

**Comment by skabashnyuk** [comment]
Can you please describe how to test this pr?

**Review by tolusha — APPROVED** [review]

**Pull request merged**